src/stan_math_backend/Transform_Mir.ml

```diff
@@ -31,18 +62,37 @@ let rec switch_expr_to_opencl available_cl_vars (Expr.Fixed.({pattern; _}) as e)
     | _, UnsizedType.(UInt | UReal) -> e
     | _, _ -> to_matrix_cl e
   in
-  let check_type args (i, t) = Expr.Typed.type_of (List.nth_exn args i) = t in
-  let any_req_met args = List.exists ~f:(List.for_all ~f:(check_type args)) in
+  let check_type args (i, ad, t) =
+    Expr.Typed.type_of (List.nth_exn args i) = t
+    && UnsizedType.autodifftype_can_convert
+         (Expr.Typed.adlevel_of (List.nth_exn args i))
+         ad
+  in
+  let any_rest_met args =
+    List.exists ~f:(List.for_all ~f:(fun t -> check_type args t))
+  in
   let maybe_map_args args req_args =
-    match any_req_met args req_args with
-    | true -> List.map args ~f:to_cl
-    | false -> args
+    match any_rest_met args req_args with
+    | true -> args
+    | false -> List.map args ~f:to_cl
+  in
+  let is_fn_opencl_supported f =
+    List.mem opencl_supported_functions
```

Comment:
Maybe we should turn `opencl_supported_functions` into a set? With like `|> String.Set.of_alist` or something.

Reply:
Done.